### ISSUE TITLE
Paragraph: Wrap with indentation

### ISSUE THREAD

**Issue by topaxi** [ratatui/ratatui#284]
## Problem

When using paragraphs and wrapping, I'd like to have an option to add some padding/indentation after the wrapping. In a way the opposite of `trim`.

## Solution

Something like this could work for me:

```rust
Paragraph::from(r#"
 - Lorem ipsum dolor sit amet, consectetur adipiscing elit.
 - Maecenas bibendum lobortis turpis, sed viverra ipsum rhoncus ac.
"#).wrap(Wrap { indent: "   " });
```

Could output something like:

```md
 - Lorem ipsum dolor sit amet, consectetur 
   adipiscing elit.
 - Maecenas bibendum lobortis turpis, sed
   viverra ipsum rhoncus ac.
```

Another use-case I have would be lines prefixed with a timestamp where I do want the wrapping of the lines to indent the length of the timestamp:

```
2023-06-29 | Lorem ipsum dolor sit amet,
             | consectetur adipiscing elit.
2023-06-29 | Maecenas bibendum lobortis
             | turpis, sed viverra ipsum
             | rhoncus ac.
```

## Alternatives

I'm not sure, I'd probably copy a lot from the wrapping logic and adjust it to the behavior I need :)

**Comment by joshka** [comment]
I really want to see some ideas for how the wrapping API should evolve to include new features. Paging @lthoerner and @TimerErTim 

I think (but might be wrong about this) that wrapping should be defined on Text objects rather than on Paragraph, which enables other widgets to not have to handle wrapping.

We have a few different requirements that should be able to be combined (and the combination of those should be considered while implementing this and other wrapping related features. So far that is:
- Where to wrap (the boundary)
- Whether to trim
- Whether to indent (and how)
- Are there more related conditions on this?

**Comment by lthoerner** [comment]
I don't think that this is related enough to wrapping, I think it's more related to `trim`. We could either make `trim` an enum rather than a bool.
```rs
enum Trim {
    TrimOnly,
    TrimAndIndent,
}
```
Then `Paragraph.trim` would be an `Option<Trim>`.

**Comment by topaxi** [comment]
I agree, that indeed makes more sense!

**Comment by joshka** [comment]
It's worth taking a look at the original reason that trim was included as it contains some of the rationale for choosing to have it on the Wrap class as a bool. I somewhat disagree with the logic (as enums generally can describe something a lot better), though I've only briefly scanned the thread.
- https://github.com/fdehau/tui-rs/pull/327

**Comment by TimerErTim** [comment]
So I am getting a little bit into implementation details here:

I think considering everything else, we should implement the wrapping methods on `Line`. That is the most low level implementation, where the "settings" are given as parameters. F.e.:

```rs
impl Line {
    fn wrap(self, width: u16, trim: Trim, indentation: Indentation) -> Vec<Line> // or Iter<Line>
    
    fn truncate(self, width: u16) -> Vec<Line> // or Iter<Line>
    
    ....
}
```

Then we could make an enum describing the different line composing methods:

```rs
enum TextComposition {
    WordWrapping(Trim, Indentation),
    CharWrapping(Trim, Indentation),
    Truncation,
}
```

and use it to wrap `Text`:

```rs
impl TextComposition {
    fn compose(&self, line: Line, width: u16) -> Vec<Line> {
        match self {
           TextComposition::Truncation -> line.truncate(width),
           TextComposition::WordWrapping(trim, indentation) -> line.wrap(width, trim, indentation),
           ...
        }
    }
}
```

and to make things easier, implement this on any `IntoIterator<Line>`:

```rs
impl<T: IntoIterator<Line> T {
    fn wrap(self, width: u16, composition: TextComposition) -> Iter<Line> {
        self.into_iter().flat_map(|line| composition.compose(text, width))
    }
}
```


Maybe I am missing something here, but adding new or expanding existing compositing methods would only require changing or adding enum entries and low level compositing methods on `Line` respectively. This is also way more straightforward than the current approach. 

Any feedback?

**Comment by joshka** [comment]
My biggest feedback is to start with explaining the sum total of options that will be supported around wrapping. Define them concisesly Right now I know that there are a few pieces that interact and a few suggestions, but I don't know how they should interact. Without that any feedback comes from a place where we have different understanding of this.

**Comment by TimerErTim** [comment]
So with my current understanding, `Paragraph` right now is the only widget that makes use of the `LineComposer`s. 
They are used to implement different line fitting behaviors (Truncating too long lines, wrapping them into multiple ones) in case the `Paragraph` is too slim in order to render whole lines. 

All `LineComposer`s depend on the max allowed width (the width of the actual `Paragraph` widget).

Currently there are:
- Truncation: Simply truncates the line and discards the parts that would overflow the `Paragraph`. Depending on the lines alignment, it truncates from the front or from the end.
  - Depends on: Alignment of `Line`; allowed witdth
- WordWrap: Wraps the line into multiple ones, so that each line fits into the given width, while trying to preserve whole words.
  - Depends on: allowed width
  - Configurable:
    - Trim: Whether or not to remove leading whitespaces of resulting lines

Currently existing suggestions:
- WordWrap: Wraps the line into multiple ones, so that each line fits into the given width, while trying to preserve whole words.
  - Depends on: allowed width
  - Configurable:
    - Trim: Whether or not to remove leading whitespaces of resulting lines
    - Indentation: Prefix every resulting line with a specific string (not sure if this is appropriate for the composers. Should maybe be done as preprocessing on the `Paragraph`s `Text`)
- CharWrap: Wraps the line into multiple ones, so that each line fits into the given width, while not trying to preserve whole words.
  - Depends on: allowed width
  - Configurable:
    - Trim: Whether or not to remove leading whitespaces of resulting lines

Currently `LineComposer` implementations are structs that own a specific configuration and the lines to wrap in form of an iteratable  during creation  (`LineComposer::new`). They expose a method to retrieve the next whole composed/processed/wrapped line. 

The `Paragraph` widget makes use of this by (depending on its own config) creating a certain `LineComposer` with the content of the `Paragraph`s `Text`. It then retrieves the composed lines one after another and either discards them (if they are out of bounds given the defined offset) or renders them onto the terminal.

That has proven to be very hard to maintain because of Iterator and Life Cycle shenanigans.

**Comment by lthoerner** [comment]
> I think considering everything else, we should implement the wrapping methods on Line. That is the most low level implementation, where the "settings" are given as parameters.

See #259.

**Comment by TimerErTim** [comment]
> > I think considering everything else, we should implement the wrapping methods on Line. That is the most low level implementation, where the "settings" are given as parameters.
> 
> See #259.

Oh well sorry, forgot there already was a PR for that

**Comment by topaxi** [comment]
I found this via googling which solves my personal use-cases above:
https://gist.github.com/sunshowers/457102aabd9db3bb23d61f2b4259bdd0

It works directly on Line or Text, I slightly modified it to also work on a slice of spans and `initial_indent` and `subsequent_indent` to be slices of `Span`s too, to cover my needs.

**Comment by joshka** [comment]
Yup, this was added over in https://github.com/ratatui-org/ratatui/pull/193#issuecomment-1565957104 a long while back. It's very close to what I'd see the wrapping to look like inside ratatui in the future (and it's what I've had in mind when discussing this with various devs), but there's a missing piece that the wrapping should meet all the requirements of the wrapping RFC, which is only partially designed. It's been a while since I read that code, so it might meet many of the features already.

**Comment by joshka** [comment]
Closing for now - tracking in #293 and the related discussion thread https://github.com/ratatui/ratatui/discussions/1925
